pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SingleSnapshotAbortedTxnProcessorImpl.java

```diff
@@ -87,7 +88,16 @@ public CompletableFuture<PositionImpl> recoverFromSnapshot() {
                     PositionImpl startReadCursorPosition = null;
                     try {
                         while (reader.hasMoreEvents()) {
-                            Message<TransactionBufferSnapshot> message = reader.readNext();
+                            Message<TransactionBufferSnapshot> message;
+                            try {
+                                message = reader.readNextAsync().get(30, TimeUnit.SECONDS);
+                            } catch (Exception ex) {
+                                Throwable t = FutureUtil.unwrapCompletionException(ex);
```

Comment:
please wrap TimeoutException, otherwise, tc will not retry the end tb op and client will not reconnect
https://github.com/apache/pulsar/blob/1be5a69a079594d8d96d2a0ab7ab8b389da8865e/pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java#L182-L187

please check the return error code, current return error code seem not correct.
https://github.com/apache/pulsar/blob/1be5a69a079594d8d96d2a0ab7ab8b389da8865e/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java#L1344-L1352

Reply:
Already fixed.